lib/Sema/TypeCheckStmt.cpp

```diff
@@ -449,8 +449,25 @@ class StmtChecker : public StmtVisitor<StmtChecker, Stmt*> {
     Type ResultTy = TheFunc->getBodyResultType();
     if (!ResultTy || ResultTy->hasError())
       return nullptr;
+    
+    Expr *E = RS->getResult();
 
     if (!RS->hasResult()) {
+
+      // If the return statement in a failable initializer is missing a 'nil',
+      // show an error and provide a fix
+      if (auto ctor = dyn_cast_or_null<ConstructorDecl>(
+                                          TheFunc->getAbstractFunctionDecl())) {
+        if (ctor->isFailable()) {                                    
```

Comment:
@HassanElDesouky As I mentioned in the last comment on the issue, I think this is already correct...
Because when a `ReturnStmt` it is a `return` or an implicit return, so this would be also valid inside a failable init.
Here an example:
```swift
init?(label: String?) { 
         guard let actualLabel = label else {
             return nil
         }

         self.label = actualLabel
         return // You could add an explicit return
     }
```
```swift
init?(label: String?) { 
         guard let actualLabel = label else {
             return nil
         }

         self.label = actualLabel
     } // There is an implicit return statement here I think
```
In both examples the code is valid :)
So in my opinion, if the incorrect message from the SR is fixed on master and this is already fixed. And you could just add the test case here 
cc @xedin

Reply:
I'm sorry I don't quite understand. So you are suggesting to just remove all of this and a test case? 
If so... which diagnostic this should be diagnosed to?

Reply:
@LucianoPAlmeida I don't have a master branch, the master branch isn't compiling on Ubuntu. You can check the error message [here](https://bugs.swift.org/browse/SR-12248?focusedCommentId=55348&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-55348) Therefore, I'm on an older branch.

Reply:
Sorry, just edited the previous comment, but what I mean is that I think this code is already correct if it compiles and it shouldn't be diagnosed because is valid code like the examples. So it shouldn't be diagnosed.
And if this code compiles on master, I think the issue is already fixed :)

Reply:
I understand now, thank you! However, what do you mean by "you could just add the test case here"?

Reply:
Yeah, but come to think about it... just note that when I say compile correctly I mean typecheck correctly, but should fail on SIL verify (Definitive Initialization)... because label member is not initialized in all paths ... so although the actual message seems confusing, it is correct, label member is not initialized on the return early guard case ...

Reply:
Yeah, I agree with @LucianoPAlmeida. This is something that should be handled in SIL diagnostics and not the typechecker. Anyway, this case is already handled by DefiniteInitialization:

```swift
// swiftc -emit-sil
/Users/suyashsrijan/Desktop/test.swift:3:5: error: constant 'self.label' used before being initialized
    init?(label: String?) {
    ^
```

as described in the JIRA ticket, but I don't think we can assume that every `return` must be a `return nil` (as shown in the examples above), so it might be a bit tricky to detect the right "exits"  where a `return nil` is the right answer.

Reply:
I would argue that in a failable initialiser any `return` before all properties are initialised should be an error (at the `return`, not at the `init`, as that is where the error actually is). I would also argue that the compiler should provide a fix-it to convert it to a `return nil` as the user forgetting the `nil` part is the most likely cause of the error (and is something the current error message and location don't actually explain in any way)

Reply:
> so it might be a bit tricky to detect the right "exits" where a return nil is the right answer

Agree @theblixguy, I was thinking in something like tweak the wording of diagnostic message for cases like this to something like `'self.label' should be initialized on all paths/branches...`, when there is no explicit reference to the member in at least one of the paths ... maybe this could make it more clear.

Reply:
> I would also argue that the compiler should provide a fix-it to convert it to a return nil as the user forgetting the nil part is the most likely cause of the error

Humm, since in failable inits that's the most commom way people use `return` it makes sense

Reply:
I agree with @pilky. It seems like the problem here is that `return` without `nil` should be diagnosed in situations when it's used before all of the properties are initialized. Which seems like it belongs in SIL diagnostics.